pkgs/top-level/named-hierarchy-files.nix

```diff
@@ -0,0 +1,108 @@
+/*
+Internal Nixpkgs file, it may be removed or changed at any point without warning
+
+This file contains the mapping from attribute name to the file in pkgs/by-name that contains a definition of it.
+No validity checks are performed, instead this file is optimized for performance.
+TODO: Link to manual and validity checks
+
+Example: If `pkgs/by-name/he/hello` exists, the result will have a `hello = pkgs/by-name/he/hello/package.nix` entry.
+
+Type: AttrsOf Path
+*/
+
+# Because the result might be fairly large, we only want compute it once.
+# In order to ensure this, we make use of Nix's `import` cache.
+# This cache memoizes the resulting value of the imported file, notably including any `let in` bindings at the top of the file.
+# This is why everything here is computed in a `let in` binding at the top.
```

Comment:
This could be moved to a test. The test is as simple as `isAttrs (import <this file>)`.